internal/plugin6/grpc_provider.go

```diff
@@ -1489,22 +1500,255 @@ func (p *GRPCProvider) ConfigureStateStore(r providers.ConfigureStateStoreReques
+	}
+
+	if resp.Diagnostics.HasErrors() {
+		logger.Trace("GRPCProvider.v6: ReadStateBytes: experienced an error when receiving state data from the provider", resp.Diagnostics.Err())
+		return resp
+	}
+
+	if buf.Len() != expectedTotalLength {
+		logger.Trace("GRPCProvider.v6: ReadStateBytes: received %d bytes but expected the total bytes to be %d.", buf.Len(), expectedTotalLength)
+
+		err = fmt.Errorf("expected state file of total %d bytes, received %d bytes",
+			expectedTotalLength, buf.Len())
+		resp.Diagnostics = resp.Diagnostics.Append(err)
+		return resp
+	}
+
+	// We're done, so close the stream
+	logger.Trace("GRPCProvider.v6: ReadStateBytes: received all chunks, total bytes: ", buf.Len())
+	err = client.CloseSend()
+	if err != nil {
+		resp.Diagnostics = resp.Diagnostics.Append(grpcErr(err))
+		return resp
+	}
+
+	// Add the state data in the response once we know there are no errors
+	resp.Bytes = buf.Bytes()
+
+	return resp
+}
+
+func (p *GRPCProvider) WriteStateBytes(r providers.WriteStateBytesRequest) (resp providers.WriteStateBytesResponse) {
+	logger.Trace("GRPCProvider.v6: WriteStateBytes")
+
+	// WriteStateBytes can be more sensitive to message sizes
+	// so we ensure it aligns with (the lower) terraform-plugin-go.
+	opts := grpc.MaxSendMsgSizeCallOption{
+		MaxSendMsgSize: grpcMaxMessageSize,
+	}
+
+	chunkSize, ok := p.stateChunkSize[r.TypeName]
+	if !ok {
+		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("Unable to determine chunk size for provider %s; this is a bug in Terraform - please report it", r.TypeName))
+		return resp
+	}
+
+	schema := p.GetProviderSchema()
+	if schema.Diagnostics.HasErrors() {
+		resp.Diagnostics = schema.Diagnostics
+		return resp
+	}
+
+	if _, ok := schema.StateStores[r.TypeName]; !ok {
+		resp.Diagnostics = resp.Diagnostics.Append(fmt.Errorf("unknown state store type %q", r.TypeName))
+		return resp
+	}
+
+	// Start the streaming RPC with a context. The context will be cancelled
+	// when this function returns, which will stop the stream if it is still
+	// running.
+	ctx, cancel := context.WithCancel(p.ctx)
+	defer cancel()
+
+	client, err := p.client.WriteStateBytes(ctx, opts)
+	if err != nil {
+		resp.Diagnostics = resp.Diagnostics.Append(grpcErr(err))
+		return resp
+	}
+
+	buf := bytes.NewBuffer(r.Bytes)
+	var totalLength int64 = int64(len(r.Bytes))
+	var totalBytesProcessed int
+	for {
+		chunk := buf.Next(chunkSize)
+
+		if len(chunk) == 0 {
+			// The previous iteration read the last of the data. Now we finish up.
+			protoResp, err := client.CloseAndRecv()
+			if err != nil {
+				resp.Diagnostics = resp.Diagnostics.Append(grpcErr(err))
+				return resp
+			}
+			resp.Diagnostics = resp.Diagnostics.Append(convert.ProtoToDiagnostics(protoResp.Diagnostics))
+			if resp.Diagnostics.HasErrors() {
+				return resp
+			}
+			break
+		}
+
+		var meta *proto6.RequestChunkMeta
+		if totalBytesProcessed == 0 {
+			// Send metadata with the first chunk only
```

Comment:
This could be covered by our tests, as the mock can throw errors if the request differs from what the mock expects.

Reply:
Ah yes, the updated test does. Thanks!